pilot/pkg/networking/core/v1alpha3/envoyfilter.go

```diff
@@ -301,3 +301,89 @@ func insertNetworkFilter(listenerName string, filterChain *listener.FilterChain,
 	log.Infof("EnvoyFilters: Rebuilt network filter stack for listener %s (from %d filters to %d filters)",
 		listenerName, oldLen, len(filterChain.Filters))
 }
+
+func buildListenerFromEnvoyConfig(value *types.Value) (*xdsapi.Listener, error) {
+	listener := xdsapi.Listener{}
+	val := value.GetStringValue()
+	if val != "" {
+		jsonum := &jsonpb.Unmarshaler{}
+		r := strings.NewReader(val)
+		err := jsonum.Unmarshal(r, &listener)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &listener, nil
+}
+
+func applyConfigPatches(listeners []*xdsapi.Listener, env *model.Environment, labels model.LabelsCollection) []*xdsapi.Listener {
+	filterCRD := getUserFiltersForWorkload(env, labels)
+	if filterCRD == nil {
+		return listeners
+	}
+
+	for _, cp := range filterCRD.ConfigPatches {
+		if cp.GetApplyTo() != networking.EnvoyFilter_LISTENER {
+			continue
+		}
+
+		if cp.GetMatch() == nil {
+			newListener, err := buildListenerFromEnvoyConfig(cp.GetPatch().GetValue())
+			if err != nil {
+				log.Warnf("Failed to unmarshal provided value into listener")
+				return listeners
+			}
+
+			listeners = append(listeners, newListener)
+		}
+	}
+
+	for _, cp := range filterCRD.ConfigPatches {
```

Comment:
move this for loop logic into the for loop above. Otherwise, the current logic will apply patches into the newly inserted listeners as well, which is not what the user expects.